File: 프로그래머스/3/118669. 등산코스 정하기/등산코스 정하기.cpp
#include <iostream>
#include <string>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

int N;
vector<pair<int,int>> adjList[50001];

int check[50001];
int dist[50001];
int visited[50001];

vector<int> solution(int n, vector<vector<int>> paths, vector<int> gates, vector<int> summits) {
    vector<int> answer;

    for(int i = 0; i < gates.size(); ++i){
        check[gates[i]] = 1;
    }
    for(int i = 0; i < summits.size(); ++i){
        check[summits[i]] = 2;
    }
    
    // N, 간선 추가
    N = n;
    for(int i = 0; i < paths.size(); ++i){
        adjList[paths[i][0]].push_back({paths[i][1], paths[i][2]});
        adjList[paths[i][1]].push_back({paths[i][0], paths[i][2]});
    }
    
    int minDist = 2100000000;
    int start = 50002, end = 0;
    
    for(int i = 0; i < gates.size(); ++i){
        int gate = gates[i];
        
        priority_queue<pair<int,int>> pq;
        
        pq.push({0, gate});
        
        visited[gate] = gate;
        dist[gate] = 0;
        
        // cout << "게이트 탐색 : " << gate << '\n';
        
        while(!pq.empty()){
            int d = -pq.top().first;
            int cur = pq.top().second;
            pq.pop();
            
            if(d > minDist) continue;
            
            if(check[cur] == 2){
                if(d < minDist) {
                    minDist = d;
                    start = gate;
                    end = cur;
                }else if(d == minDist && cur < end){
                    start = gate;
                    end = cur;
                }
                
                continue;
            }
        
            
            for(int j = 0; j < adjList[cur].size(); ++j) {
                int nx = adjList[cur][j].first;
                int nd = max(d, adjList[cur][j].second);
                
                if(check[nx] == 1) continue;
                if(visited[nx] == gate && dist[nx] <= nd) continue;
                
                visited[nx] = gate;
                dist[nx] = nd;
                
                pq.push({-nd, nx});
                // cout << nx << " " << nd << '\n';
            }
        }
    }

    answer.push_back(end);
    answer.push_back(minDist);
    
    return answer;
}

File: 프로그래머스/3/118669. 등산코스 정하기/등산코스 정하기.java
import java.util.*;

class Solution {

    static Set<Integer> entr;
    static Set<Integer> smt;
    static List<int[]>[] adj;
    static int[] visited;
    static int[] intens;
    
    public int[] solution(int n, int[][] paths, int[] gates, int[] summits) {
        int[] answer = {0, 0};
        
        entr = new HashSet<>();
        smt = new HashSet<>();
        adj = new ArrayList[n+1];
        visited = new int[n+1];
        intens = new int[n+1];
        
        for(int i = 0; i < n+1; ++i) adj[i] = new ArrayList<>();
        for(int i = 0; i < gates.length; ++i) entr.add(gates[i]);
        for(int i = 0; i < summits.length; ++i) smt.add(summits[i]);
        for(int i = 0; i < paths.length; ++i){
            adj[paths[i][0]].add(new int[] {paths[i][1], paths[i][2]});
            adj[paths[i][1]].add(new int[] {paths[i][0], paths[i][2]});
        }
        
        int minIntensity = 10000001;
        int start = 50001, end = 0;
        
        for(int i = 0; i < gates.length; ++i) {
            int gate = gates[i];
            
            PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
            pq.add(new int[] {0, gate});
            visited[gate] = gate;
            intens[gate] = 0;
            
            while(!pq.isEmpty()) {
                int inten = pq.peek()[0];
                int cur = pq.peek()[1];
                pq.poll();
                
                if(inten > minIntensity) continue;
                
                if(smt.contains(cur)) {
                    if(inten < minIntensity || (inten == minIntensity && cur < end)) {
                        minIntensity = inten;
                        start = gate;
                        end = cur;
                    }
                    continue;
                }
                
                for(int j = 0; j < adj[cur].size(); ++j) {
                    int nx = adj[cur].get(j)[0];
                    int ni = Math.max(inten, adj[cur].get(j)[1]);

                    if(entr.contains(nx)) continue;
                    if(visited[nx] == gate && intens[nx] <= ni) continue;

                    visited[nx] = gate;
                    intens[nx] = ni;

                    pq.add(new int[] {ni, nx});
                }
            }
        }
        
        answer[0] = end;
        answer[1] = minIntensity;
        
        return answer;
    }
}